framework/src/main/java/com/github/dbmdz/flusswerk/framework/rabbitmq/RabbitClient.java

```diff
@@ -130,17 +132,59 @@ byte[] serialize(Message message) throws IOException {
     return objectMapper.writeValueAsBytes(message);
   }
 
-  public void ack(Envelope envelope) {
+  public void ack(com.github.dbmdz.flusswerk.framework.model.Envelope envelope) {
+    ack(envelope.getDeliveryTag());
+  }
+
+  public void ack(long deliveryTag) {
     // The channel might not be available or become unavailable due to a connection error. In this
     // case, we wait until the connection becomes available again.
     while (true) {
       if (channelAvailable) {
         try {
-          channel.basicAck(envelope.getDeliveryTag(), SINGLE_MESSAGE);
+          channel.basicAck(deliveryTag, SINGLE_MESSAGE);
```

Comment:
Je nachdem, wie aufwändig das ist, aber hier ist viel doppelter Code drinnen, der quasi die Form:

```java
while (true) {
      if (channelAvailable) {
        try {
          // Only this part is is function-specific!
          break;
        } catch (IOException | AlreadyClosedException e) {
          // Channel-level exceptions are not recoverable
          if (e instanceof AlreadyClosedException && !((AlreadyClosedException) e).isHardError()) {
            recoverChannel();
          } else {
            log.warn(
                "Failed to reject message from RabbitMQ: '{}', waiting for channel to become available again",
                e.getMessage());
            channelAvailable = false;
          }
        }
      }
      // We loop here because the signal might be triggered due to what the JVM documentation calls
      // a 'spurious wakeup', i.e. the signal is triggered even though no connection recovery has
      // yet happened.
      while (!channelAvailable) {
        channelLock.lock();
        channelAvailableAgain.awaitUninterruptibly();
        channelLock.unlock();
      }
    }
```

hat.

D.h. eigentlich ändert sich nur im `try`Teil etwas. Eventuell bekommt man das mit einer Funktion hin, die als Argument zum Beispiel `operation`, damit folgende Fälle abgedeckt sind:

* `basicPublish`: `channel.basicPublish(exchange, routingKey, properties, data);`
* `basicAck`: `channel.basicAck(deliveryTag, SINGLE_MESSAGE);`
* `basicReject`: `channel.basicReject(envelope.getDeliveryTag(), SINGLE_MESSAGE);`
* `basicGet`: `response = channel.basicGet(queueName, autoAck);`
* `basicResume`: `channel.basicConsume(consumer.getInputQueue(), autoAck, consumer);`
* `basicNack`: channel.basicNack(deliveryTag, multiple, requeue);`

Wobei `basicGet` eine Rückgabe braucht.

Was meint ihr, @schmika und @bitzl :thinking:

Reply:
Ja, weniger Wiederholung wäre gut. Mir ist nur gerade noch nicht klar, mit welchem Muster man das angesichts der unterschiedlichen Parameterzahl und -typen und Rückgabewerte am besten lösen könnte...
Ich dachte an eine Methode `callRecoverable`, die ein Functional Interface und die jeweiligen Parameter für die aufzurufende Funktion als Parameter hat.
Functional Interfaces geben aber z.B. Parameterzahl und Rückgabe vor und ich weiß nicht, wie ich verschiedene Parametertypen mit varargs am besten abbilden könnte.
Fallen euch bessere Lösungen ein?